### PR TITLE
[review] Add multiprocessing.Pool

### PR DESCRIPTION
**Pull request by alexjurkiewicz**
I'm not sure how to test stub files (does mypy download them automatically? I couldn't see anything in `mypy --help` about using stubs) so this is untested, but I'm aiming to fix this error:

```
$ cat test.py
import multiprocessing
multiprocessing.Pool()
$ mypy test.py
test.py:2: error: "module" has no attribute "Pool"
```

### CODE REVIEW

**Comment by matthiaskramm** [comment]
You can test stubs from within typeshed, usually by doing something like

```
python3 tests/mypy_test.py
python tests/pytype_test.py
```

**Comment by alexjurkiewicz** [comment]
fixed tests, thanks!

File: stdlib/3/multiprocessing/__init__.pyi [hunk 1]
```diff
@@ -1,13 +1,44 @@
...
     def release(self) -> None: ...
 
+class AsyncResult():
+    def get(self, timeout: float = -1) -> Any: ...
+    def wait(self, timeout: float = -1) -> None: ...
+    def ready(self) -> bool: ...
+    def successful(self) -> bool: ...
+
+class Pool():
+    def __init__(self, processes: int = ...) -> None: ...
```
Comment (gvanrossum) [inline stdlib/3/multiprocessing/__init__.pyi]: According to the docs this has a lot more arguments. Also, it is documented that `processes = None` has a special meaning, so the type should really be `Optional[int]`.

**Comment by gvanrossum** [comment]
@alexjurkiewicz Can you please respond to the code review?

**Comment by alexjurkiewicz** [comment]
Sorry on hols! Next weekend I will be back at a computer with git to make
these changes

**Comment by alexjurkiewicz** [comment]
Thanks for waiting. I've made the requested changes. Two points:

1) the `error_callback` arguments accept a callable that takes an Exception as the sole argument. I'm not sure how to model this in Python type annotations.

2) There's another Pool definition in `pool.py` in the same directory as the file I'm editing. Should it still exist?

**Comment by gvanrossum** [comment]
> 1) the error_callback arguments accept a callable that takes an Exception as the sole argument. I'm not sure how to model this in Python type annotations.

Probably `Callable[[BaseException], None]`? Or did you mean it takes an exception class object? Then `Callable[[Type[BaseException]], None]`.

> 2) There's another Pool definition in pool.py in the same directory as the file I'm editing. Should it still exist?

Hm... The interface here is weird, and changed around Python 3.4. I suppose the idea is that you always use the functions in the multiprocessing module, not in its submodules. So perhaps it can be deleted. But I'd leave it alone for now. (I'm actually no expert on this module, and it has a lot of documentation to read through.)

> 3) multiprocessing.Pool.__init__'s initializer argument is a callable that takes initargs as arguments. I don't know how to model this either.

I think `Callable[..., None]` (or perhaps `Callable[..., Any]`?) is the way to go for the initializer, and then initargs should probably be `Sequence[Any] = ...` (leaving the actual default value unspecified by the stub).

> 4) Can the multiprocessing.Pool.__init__'s context argument definition be improved?

I don't think of context as a callable -- it looks more like it's an instance of multiprocessing.context.BaseContext (or None, indeed).

**Comment by alexjurkiewicz** [comment]
> Probably Callable[[BaseException], None]? Or did you mean it takes an exception class object? Then Callable[[Type[BaseException]], None].

Thanks, pretty sure the former is correct.

> I think `Callable[..., None]` (or perhaps `Callable[..., Any]`?) is the way to go for the initializer

The return type is ignored -- so I'm not sure if we should use None or Any. None seems stricter, which is why I picked it since if it's a problem the definition can be relaxed later (reverse would be harder).

> then initargs should probably be `Sequence[Any] = ...`

This might be a dumb question since I don't really  understand how `*` works, but would `Iterable[Any] = ...` be better?

> I don't think of context as a callable -- it looks more like it's an instance of multiprocessing.context.BaseContext (or None, indeed).

Unfortunately there is no stub for `multiprocessing.context.BaseContext`, and I'm sorry but don't have the motivation to write one because it looks very complex. I guess I could write an empty stub for it though... would you prefer that? For now, I didn't change from my original guess.

File: stdlib/3/multiprocessing/__init__.pyi [hunk 2]
```diff
@@ -1,13 +1,67 @@
 # Stubs for multiprocessing
 
-from typing import Any, Callable, Iterable, Mapping
+from typing import Any, Callable, Iterable, Mapping, Sequence, Optional, \
+    Dict, List
```
Comment (gvanrossum) [inline stdlib/3/multiprocessing/__init__.pyi]: Oh, this should use () instead of \ for the continuation line.

**Comment by alexjurkiewicz** [comment]
Updated. (btw, rather than replacing the commit each update, would you prefer incremental commits that I squash-rebase at the end?)

> Don't add an empty stub for context, just make the context argument's type Any.

Just checking, this means `Optional[Callable[Any, Any]]`, right?

**Comment by gvanrossum** [comment]
I still don't know how the context is used. Personally I'd just set it to
`context: Any = ...` .

And yes, please send me incremental commits (only push when you're ready
for a review). There's no need for you to squash at the end, I'll do that
when I merge (GitHub added that as a new feature a few months ago and it's
fantastic).

**Comment by alexjurkiewicz** [comment]
Cool. I'll come back to this next week.

**Comment by alexjurkiewicz** [comment]
pushed the last requested change

**Comment by gvanrossum** [comment]
Thanks!

**Pull request merged**